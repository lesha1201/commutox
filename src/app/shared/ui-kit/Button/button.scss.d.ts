export const btn: string;
export const outlined: string;
