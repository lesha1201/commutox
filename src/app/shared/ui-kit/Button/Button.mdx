---
name: Button
menu: UI Kit
---

import { Playground, Props } from 'docz';
import Margin from 'src/app/shared/docz-helpers/Margin';
import Button from './index';

# Button

`<Button />` is just html tag `<button />` with styles. All dom attrs of a button is acceptable.

## Types

<Playground>
  <div style={{ fontSize: '16px' }}>

    <Button style={{ marginRight: '1em' }}>Default</Button>

    <Button theme="outlined" style={{ marginRight: '1em' }}>
      Outlined
    </Button>

    <Button disabled style={{ marginRight: '1em' }}>
      Disabled
    </Button>

    <Button disabled theme="outlined" style={{ marginRight: '1em' }}>
      Disabled
    </Button>

  </div>
</Playground>

## Sizes

<Playground>
  <div style={{ fontSize: '16px' }}>

    <Margin type="bottom">
      <Button>Default</Button>
    </Margin>

    <Margin type="bottom">
      <Button size="wide">Wide</Button>
    </Margin>

    <Margin>
      <Button size="full">Full</Button>
    </Margin>

  </div>
</Playground>

## Colors

<Playground>
  <div style={{ fontSize: '16px' }}>

    <div>
    <h3>Default</h3>
    <Button mainColor="primary" style={{ marginRight: '1em' }}>Primary</Button>

    <Button mainColor="success" style={{ marginRight: '1em' }}>
      Success
    </Button>

    <Button mainColor="danger">
      Danger
    </Button>
    </div>

    <div>
    <h3>Outlined</h3>
    <Button mainColor="primary" theme="outlined" style={{ marginRight: '1em' }}>Primary</Button>

    <Button mainColor="success" theme="outlined" style={{ marginRight: '1em' }}>
      Success
    </Button>

    <Button mainColor="danger" theme="outlined">
      Danger
    </Button>
    </div>

  </div>
</Playground>

## "As" Prop

You can specify what element to use.

```jsx
// Using tag name
<Button as="a" href="#">Link</Button>

// Using another component (should pass className prop in order styles to be applied)
<Button as={Link} to="#">Router Link</Button>
```

## Props

<Props of={Button} />
