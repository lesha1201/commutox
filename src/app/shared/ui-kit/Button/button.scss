/* * * * * * * * *
 * Button        *
 * * * * * * * * */

@import 'src/app/shared/ui-kit/sass/variables';

@function get-disabled-color($color) {
  @return adjust-color($color, $saturation: -25%, $alpha: -0.3);
}

.common {
  display: inline-block;
  padding: 0.6em 1.3em;

  border-radius: 0.3em;
  border: 1px solid transparent;
  color: $color-primary-white;
  cursor: pointer;
  outline: none;
  text-decoration: none;
  transition: all 0.2s;

  font-size: 1em;
  font-family: inherit;
  white-space: nowrap;

  &:not(:disabled):hover,
  &:not(:disabled):focus {
    transform: translateY(-0.05em);
  }

  &:not(:disabled):active {
    transform: translateY(0.05em);
  }

  &:disabled {
    cursor: not-allowed;
  }

  /* Modifiers */
  &.full {
    width: 100%;
  }

  &.wide {
    padding: 0.6em 2.1em;
  }
}

/* Themes */
.painted {
  composes: common;

  /* Modifiers */
  &.primary {
    background-color: $color-primary;
    box-shadow: 0 0.2em 0.4em rgba($color-primary, 0.15);

    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: darken($color-primary, 1);
      box-shadow: 0 0.3em 0.6em rgba($color-primary, 0.2);
    }

    &:not(:disabled):active {
      box-shadow: 0 0.1em 0.2em rgba($color-primary, 0.2);
    }

    &:disabled {
      background-color: get-disabled-color($color-primary);
    }
  }

  &.success {
    background-color: $color-success;
    box-shadow: 0 0.2em 0.4em rgba($color-success, 0.15);

    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: darken($color-success, 1);
      box-shadow: 0 0.3em 0.6em rgba($color-success, 0.2);
    }

    &:not(:disabled):active {
      box-shadow: 0 0.1em 0.2em rgba($color-success, 0.2);
    }

    &:disabled {
      background-color: get-disabled-color($color-success);
    }
  }

  &.danger {
    background-color: $color-danger;
    box-shadow: 0 0.2em 0.4em rgba($color-danger, 0.15);

    &:not(:disabled):hover,
    &:not(:disabled):focus {
      background-color: darken($color-danger, 1);
      box-shadow: 0 0.3em 0.6em rgba($color-danger, 0.2);
    }

    &:not(:disabled):active {
      box-shadow: 0 0.1em 0.2em rgba($color-danger, 0.2);
    }

    &:disabled {
      background-color: get-disabled-color($color-danger);
    }
  }
}

.outlined {
  composes: common;

  background-color: transparent;

  &:not(:disabled):hover {
    background-color: transparent;
  }

  /* Modifiers */
  &.primary {
    border-color: $color-primary;
    color: $color-primary;

    &:not(:disabled):hover,
    &:not(:disabled):focus {
      box-shadow: 0 0.3em 0.6em rgba($color-primary, 0.1);
    }

    &:not(:disabled):active {
      box-shadow: 0 0.1em 0.2em rgba($color-primary, 0.1);
    }

    &:disabled {
      color: get-disabled-color($color-primary);
      border-color: get-disabled-color($color-primary);
    }
  }

  &.success {
    border-color: $color-success;
    color: $color-success;

    &:not(:disabled):hover,
    &:not(:disabled):focus {
      box-shadow: 0 0.3em 0.6em rgba($color-success, 0.1);
    }

    &:not(:disabled):active {
      box-shadow: 0 0.1em 0.2em rgba($color-success, 0.1);
    }

    &:disabled {
      color: get-disabled-color($color-success);
      border-color: get-disabled-color($color-success);
    }
  }

  &.danger {
    border-color: $color-danger;
    color: $color-danger;

    &:not(:disabled):hover,
    &:not(:disabled):focus {
      box-shadow: 0 0.3em 0.6em rgba($color-danger, 0.1);
    }

    &:not(:disabled):active {
      box-shadow: 0 0.1em 0.2em rgba($color-danger, 0.1);
    }

    &:disabled {
      color: get-disabled-color($color-danger);
      border-color: get-disabled-color($color-danger);
    }
  }
}
