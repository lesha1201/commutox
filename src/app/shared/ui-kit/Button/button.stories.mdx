import { Meta, Story, Props, Preview } from '@storybook/addon-docs/blocks';

import { Margin } from 'app/shared/storybook-helpers';
import Button from './index';

# Button

<Meta title="UI Kit/Button" component={Button} />

`<Button />` is just html tag `<button />` with styles. All dom attrs of a button is acceptable.

## Stories

### Types

<Preview withToolbar>
  <Story name="types">
    <div style={{ fontSize: '16px' }}>
      <Button style={{ marginRight: '1em' }}>Default</Button>
      <Button theme="outlined" style={{ marginRight: '1em' }}>
        Outlined
      </Button>
      <Button disabled style={{ marginRight: '1em' }}>
        Disabled
      </Button>
      <Button disabled theme="outlined" style={{ marginRight: '1em' }}>
        Disabled
      </Button>
    </div>
  </Story>
</Preview>

### Sizes

<Preview withToolbar>
  <Story name="sizes">
    <div style={{ fontSize: '16px' }}>
      <Margin type="bottom">
        <Button>Default</Button>
      </Margin>
      <Margin type="bottom">
        <Button size="wide">Wide</Button>
      </Margin>
      <Margin>
        <Button size="full">Full</Button>
      </Margin>
    </div>
  </Story>
</Preview>

### Colors

<Preview withToolbar>
  <Story name="colors">
    <div style={{ fontSize: '16px' }}>
      <Margin type="bottom">
        <h3 style={{ marginBottom: '8px' }}>Default</h3>
        <Button mainColor="primary" style={{ marginRight: '1em' }}>
          Primary
        </Button>
        <Button mainColor="success" style={{ marginRight: '1em' }}>
          Success
        </Button>
        <Button mainColor="danger">Danger</Button>
      </Margin>
      <div>
        <h3 style={{ marginBottom: '8px' }}>Outlined</h3>
        <Button mainColor="primary" theme="outlined" style={{ marginRight: '1em' }}>
          Primary
        </Button>
        <Button mainColor="success" theme="outlined" style={{ marginRight: '1em' }}>
          Success
        </Button>
        <Button mainColor="danger" theme="outlined">
          Danger
        </Button>
      </div>
    </div>
  </Story>
</Preview>

## "As" Prop

You can specify what element to use.

```jsx
// Using tag name
<Button as="a" href="#">Link</Button>

// Using another component (should pass className prop in order styles to be applied)
<Button as={Link} to="#">Router Link</Button>
```

## Props

Button also accepts all dom props depending on html tag (button, a, etc) it uses.

<Props of={Button} />
