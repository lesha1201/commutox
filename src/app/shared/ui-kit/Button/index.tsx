import cx from 'classnames';
import * as React from 'react';

import * as style from './button.scss';

/* Typings */
export interface IProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  theme: 'painted' | 'outlined';
  mainColor: 'primary' | 'success' | 'danger';
  size?: 'wide' | 'full';
  children?: React.ReactNode;
}

/* Component */
function Button({ children, theme, size, mainColor, className, ...domAttrs }: IProps) {
  const cn = cx(className, style[theme], style[mainColor], size && style[size]);

  return (
    <button className={cn} {...domAttrs}>
      {children}
    </button>
  );
}

Button.defaultProps = {
  theme: 'painted',
  mainColor: 'primary',
};

export default Button;
