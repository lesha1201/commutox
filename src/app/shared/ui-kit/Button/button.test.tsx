import * as React from 'react';
import { render } from 'react-testing-library';

import Button from '.';

describe('<Button />', () => {
  it('should render', () => {
    const { container } = render(<Button>button</Button>);

    expect(container.firstChild).toMatchSnapshot();
  });

  // it("should accept any button's dom attrs", () => {
  // });

  describe('themes', () => {
    it('with theme "painted"', () => {
      const { container } = render(<Button theme="painted">button</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('with theme "outlined"', () => {
      const { container } = render(<Button theme="outlined">button</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('colors', () => {
    it('with color "primary"', () => {
      const { container } = render(<Button mainColor="primary">button</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('with color "success"', () => {
      const { container } = render(<Button mainColor="success">button</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('with color "danger"', () => {
      const { container } = render(<Button mainColor="danger">button</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('sizes', () => {
    it('with size "wide"', () => {
      const { container } = render(<Button size="wide">button</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });

    it('with size "full"', () => {
      const { container } = render(<Button size="full">button</Button>);

      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
