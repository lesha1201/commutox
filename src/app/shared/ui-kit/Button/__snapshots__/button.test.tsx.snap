// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> colors with color "danger" 1`] = `
<button
  class="painted danger"
>
  button
</button>
`;

exports[`<Button /> colors with color "primary" 1`] = `
<button
  class="painted primary"
>
  button
</button>
`;

exports[`<Button /> colors with color "success" 1`] = `
<button
  class="painted success"
>
  button
</button>
`;

exports[`<Button /> should render 1`] = `
<button
  class="painted primary"
>
  button
</button>
`;

exports[`<Button /> sizes with size "full" 1`] = `
<button
  class="painted primary full"
>
  button
</button>
`;

exports[`<Button /> sizes with size "wide" 1`] = `
<button
  class="painted primary wide"
>
  button
</button>
`;

exports[`<Button /> themes with theme "outlined" 1`] = `
<button
  class="outlined primary"
>
  button
</button>
`;

exports[`<Button /> themes with theme "painted" 1`] = `
<button
  class="painted primary"
>
  button
</button>
`;
